modules/rank-eval/src/test/java/org/elasticsearch/index/rankeval/ExpectedReciprocalRankTests.java

```diff
@@ -0,0 +1,230 @@
+ * under the License.
+ */
+
+package org.elasticsearch.index.rankeval;
+
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.common.io.stream.NamedWriteableRegistry;
+import org.elasticsearch.common.text.Text;
+import org.elasticsearch.common.xcontent.ToXContent;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentFactory;
+import org.elasticsearch.common.xcontent.XContentParseException;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.common.xcontent.XContentType;
+import org.elasticsearch.common.xcontent.json.JsonXContent;
+import org.elasticsearch.index.Index;
+import org.elasticsearch.search.SearchHit;
+import org.elasticsearch.search.SearchShardTarget;
+import org.elasticsearch.test.ESTestCase;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+import static org.elasticsearch.test.EqualsHashCodeTestUtils.checkEqualsAndHashCode;
+import static org.elasticsearch.test.XContentTestUtils.insertRandomFields;
+import static org.hamcrest.CoreMatchers.containsString;
+
+public class ExpectedReciprocalRankTests extends ESTestCase {
+
+    private static final double DELTA = 10E-14;
+
+    public void testProbabilityOfRelevance() {
+        ExpectedReciprocalRank err = new ExpectedReciprocalRank(5);
+        assertEquals(0.0, err.probabilityOfRelevance(0), 0.0);
+        assertEquals(1d/32d, err.probabilityOfRelevance(1), 0.0);
+        assertEquals(3d/32d, err.probabilityOfRelevance(2), 0.0);
+        assertEquals(7d/32d, err.probabilityOfRelevance(3), 0.0);
+        assertEquals(15d/32d, err.probabilityOfRelevance(4), 0.0);
+        assertEquals(31d/32d, err.probabilityOfRelevance(5), 0.0);
+    }
+
+    /**
+     * Assuming the result ranking is
+     *
+     * <pre>{@code
+     * rank | relevance | probR / r | p        | p * probR / r
+     * -------------------------------------------------------
+     * 1    | 3         | 0.875     | 1        | 0.875       |
+     * 2    | 2         | 0.1875    | 0.125    | 0.0234375   |
+     * 3    | 0         | 0         | 0.078125 | 0           |
+     * 4    | 1         | 0.03125   | 0.078125 | 0.00244140625 |
+     * }</pre>
+     *
+     * err => sum of last column
+     */
+    public void testERRAt() {
+        List<RatedDocument> rated = new ArrayList<>();
+        int[] relevanceRatings = new int[] { 3, 2, 0, 1};
+        SearchHit[] hits = new SearchHit[relevanceRatings.length];
+        for (int i = 0; i < relevanceRatings.length; i++) {
+            rated.add(new RatedDocument("index", Integer.toString(i), relevanceRatings[i]));
+            hits[i] = new SearchHit(i, Integer.toString(i), new Text("type"), Collections.emptyMap());
+            hits[i].shard(new SearchShardTarget("testnode", new Index("index", "uuid"), 0, null));
```

Comment:
if `SearchShardTarget` is the same for all hits, would it be reasonable to create it outside of `for` loop? here and in the following test?